src/idl/src/annotation.c

```diff
@@ -510,6 +510,299 @@ annotate_must_understand(
+    return IDL_RETCODE_SEMANTIC_ERROR;
+  }
+
+  return IDL_RETCODE_OK;
+}
+
+static idl_retcode_t
+annotate_range(
+  idl_pstate_t *pstate,
+  idl_annotation_appl_t *annotation_appl,
+  idl_node_t *node)
+{
+  idl_retcode_t ret = IDL_RETCODE_OK;
+  idl_annotation_appl_param_t *par = NULL;
+
+  IDL_FOREACH(par,annotation_appl->parameters) {
+    //are the parameters sorted, or in the order of declaration of the file
+    if (ret)
+      break;
+    if (idl_strcasecmp(par->member->declarator->name->identifier, "min") == 0) {
+      ret = check_and_attach_minmax(pstate, annotation_appl, par, node, false);
+    } else if (idl_strcasecmp(par->member->declarator->name->identifier, "max") == 0) {
+      ret = check_and_attach_minmax(pstate, annotation_appl, par, node, true);
+    }
+  }
+
+  return ret;
+
+}
+
+static idl_retcode_t
+annotate_min(
+  idl_pstate_t *pstate,
+  idl_annotation_appl_t *annotation_appl,
+  idl_node_t *node)
+{
+  return check_and_attach_minmax(pstate, annotation_appl, annotation_appl->parameters, node, false);
+}
+
+static idl_retcode_t
+annotate_max(
+  idl_pstate_t *pstate,
+  idl_annotation_appl_t *annotation_appl,
+  idl_node_t *node)
+{
+  return check_and_attach_minmax(pstate, annotation_appl, annotation_appl->parameters, node, true);
+}
+
+static idl_retcode_t
+annotate_unit(
+  idl_pstate_t *pstate,
+  idl_annotation_appl_t *annotation_appl,
+  idl_node_t *node)
+{
+  const char *name = NULL;
+  idl_literal_t *literal = NULL;
+  assert(annotation_appl);
+  assert(annotation_appl->parameters);
+  literal = annotation_appl->parameters->const_expr;
+  assert(idl_type(literal) == IDL_STRING);
+  name = literal->value.str;
+
+  if (idl_is_member(node)) {
+    ((idl_member_t*)node)->unit.annotation = annotation_appl;
+    ((idl_member_t*)node)->unit.value = name;
+  } else if (idl_is_case(node)) {
+    ((idl_case_t*)node)->unit.annotation = annotation_appl;
+    ((idl_case_t*)node)->unit.value = name;
+  } else {
+    idl_error(pstate, idl_location(annotation_appl),
+      "@unit can only be assigned to members and cases");
+    return IDL_RETCODE_SEMANTIC_ERROR;
+  }
+
+
+  return IDL_RETCODE_OK;
+}
+
+static idl_retcode_t
+annotate_datarepresentation(
+  idl_pstate_t *pstate,
+  idl_annotation_appl_t *annotation_appl,
+  idl_node_t *node)
+{
+  assert(annotation_appl);
+  assert(annotation_appl->parameters);
+  idl_literal_t *literal = annotation_appl->parameters->const_expr;
+  assert(idl_type(literal) == IDL_BITMASK);
+  allowable_data_representations_t val = (allowable_data_representations_t)literal->value.uint32;  //native type of datarepresentation is uint32_t
+
+  if (idl_requires_xtypes_functionality(node))
+    val &= ~((allowable_data_representations_t)IDL_DATAREPRESENTATION_FLAG_XCDR1);
+
+  if (0 == val) {
+    idl_error(pstate, idl_location(annotation_appl),
+      "no viable datarepresentations for type %s", idl_name(node)->identifier);
+    return IDL_RETCODE_SEMANTIC_ERROR;
+  }
+
```

Comment:
If the min/max annotations need to check for the field having been annotated before already, shouldn't this one do so, too?

Reply:
The way I understand the parser works, this is caught before the annotation contents are parsed in the annotation.c:dedup function and you will get a "Incompatible reapplication of @$name_of_annotation$" error.